File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter12/ElPasoHighSchool/ElPasoHighSchool/StudentBody.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace ElPasoHighSchool
{
    public class StudentBody : INotifyPropertyChanged
    {
        string school;
        ObservableCollection<Student> students = new ObservableCollection<Student>();

        public event PropertyChangedEventHandler PropertyChanged;

        public string School
        {
            set { SetProperty<string>(ref school, value); }
            get { return school; }
        }

        public ObservableCollection<Student> Students
        {
            set { SetProperty<ObservableCollection<Student>>(ref students, value); }
            get { return students; }
        }

        protected bool SetProperty<T>(ref T storage, T value,
                              [CallerMemberName] string propertyName = null)
        {
            if (object.Equals(storage, value))
                return false;

            storage = value;
            OnPropertyChanged(propertyName);
            return true;
        }

        protected void OnPropertyChanged(string propertyName)
        {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter12/ElPasoHighSchool/ElPasoHighSchool/StudentBody.cpp
#include "pch.h"
#include "StudentBody.h"

using namespace ElPasoHighSchool;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;

StudentBody::StudentBody()
{
    this->Students = ref new Vector<Student^>();
}

void StudentBody::School::set(String^ value)
{
    if (school != value)
    {
        school = value;
        OnPropertyChanged("School");
    }
}

void StudentBody::Students::set(IVector<Student^>^ value)
{
    if (students != value)
    {
        students = value;
        OnPropertyChanged("Students");
    }
}
